src/OpenTelemetry/Internal/BatchExportTaskWorker.cs

```diff
@@ -0,0 +1,256 @@
+    {
+    }
+
+    /// <inheritdoc/>
+    public override void Start()
+    {
+        this.workerTask = Task.Factory.StartNew(
+            this.ExporterProcAsync,
+            CancellationToken.None,
+            TaskCreationOptions.LongRunning,
+            TaskScheduler.Default).Unwrap();
+    }
+
+    /// <inheritdoc/>
+    public override bool TriggerExport()
+    {
+        if (this.cancellationTokenSource.IsCancellationRequested)
+        {
+            return false;
+        }
+
+        try
+        {
+            this.exportTrigger.Release();
+            return true;
+        }
+        catch (ObjectDisposedException)
+        {
+            return false;
+        }
+        catch (SemaphoreFullException)
+        {
+            // Semaphore is already signaled, export is pending
+            return true;
+        }
+    }
+
+    /// <inheritdoc/>
+    public override bool WaitForExport(int timeoutMilliseconds)
+    {
+        var tail = this.circularBuffer.RemovedCount;
+        var head = this.circularBuffer.AddedCount;
+
+        if (head == tail)
+        {
+            return true; // nothing to flush
+        }
+
+        if (!this.TriggerExport())
+        {
+            return false;
+        }
+
+        if (timeoutMilliseconds == 0)
+        {
+            return false;
+        }
+
+        return this.WaitForExportAsync(timeoutMilliseconds, head).GetAwaiter().GetResult();
+    }
+
+    /// <inheritdoc/>
+    public override bool Shutdown(int timeoutMilliseconds)
+    {
+        this.SetShutdownDrainTarget(this.circularBuffer.AddedCount);
+
+        try
+        {
+            this.cancellationTokenSource.Cancel();
+        }
+        catch (ObjectDisposedException)
+        {
+            return false;
+        }
+
+        if (this.workerTask == null)
+        {
+            return true;
+        }
+
+        if (timeoutMilliseconds == Timeout.Infinite)
+        {
+            this.workerTask.Wait();
+            return true;
+        }
+
+        if (timeoutMilliseconds == 0)
+        {
+            return true;
+        }
+
+        return this.workerTask.Wait(timeoutMilliseconds);
+    }
+
+    /// <inheritdoc/>
+    protected override void Dispose(bool disposing)
+    {
+        base.Dispose(disposing);
+
+        if (!this.disposed)
+        {
+            this.disposed = true;
+
+            this.cancellationTokenSource.Dispose();
+            this.exportTrigger.Dispose();
+        }
+    }
+
+    private async Task<bool> WaitForExportAsync(int timeoutMilliseconds, long targetHead)
+    {
+        var sw = timeoutMilliseconds == Timeout.Infinite
+            ? null
+            : Stopwatch.StartNew();
+
+        // There is a chance that the export task finished processing all the data from the queue,
+        // and signaled before we enter wait here, use polling to prevent being blocked indefinitely.
+        const int pollingMilliseconds = 1000;
```

Comment:
Is it possible to factor this in terms of `TimeSpan` (e.g. with `Stopwatch.Elapsed`) rather than integer milliseconds? It might make the code more readable, and I try to avoid `int` for timeouts where possible so bugs from things like mixing seconds and milliseconds don't occur.

Reply:
OK no problem - I thought it might have been tricky with the various TFMs and possible missing overloads anyway, so if there's no easy way to refactor it then let's leave it in terms of `int`.